### PR TITLE
perlPackages: add meta.mainProgram to many packages

### PR DESCRIPTION
**Pull request by malob**
###### Description of changes

Add `meta.mainProgram` for packages in `perlPackages` where the package provides a single executable and that executable's name differs from the package's `name` or `pname`, for all packages available on `{aarch64,x86_64}-darwin` and `x86_64-linux`. Also add `meta.mainProgram` for packages with multiple executables where none of the executables match the package's `name` or `pname`, and one of the executables is the obvious `mainProgram`.

This will allow `nix run` to work with 160 additional packages.

If a package I touched was missing key `meta` attributes like, `license` or `description`, I added them when the information was easy to find. I also fixed some indentation issues, and made a few other minor formatting tweaks/fixes.  

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by malob** [comment]
Result of `nixpkgs-review pr 172413` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)

File: pkgs/development/perl-modules/strip-nondeterminism/default.nix [hunk 1]
```diff
@@ -36,5 +36,6 @@ buildPerlPackage rec {
     homepage = "https://reproducible-builds.org/";
     license = licenses.gpl3;
     maintainers = with maintainers; [ pSub ];
+    mainProgram = "strip-nondeterminism";
```
Comment (SuperSandro2000) [inline pkgs/development/perl-modules/strip-nondeterminism/default.nix:39]: That shouldn't be necessary.

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]
Please remove all changes where `meta.mainProgram = pname`.

File: pkgs/development/perl-modules/strip-nondeterminism/default.nix (hunk 1 above)
Comment (malob) [reply pkgs/development/perl-modules/strip-nondeterminism/default.nix:39]: Unfortunately `buildPerlPackages` seems to mess with `pname`, so `nix run` basically doesn't work on anything in `perlPackages` by default :(

```console
~/C/nixpkgs on perlPackages-mainProgram
❯ nix eval -f . perlPackages.strip-nondeterminism.pname
"perl5.34.1-strip-nondeterminism"
```

**Comment by malob** [comment]
@stigtsp, @zakame, any concerns here?

**Review by zakame — COMMENTED** [review]
Thanks @malob! On a cursory review it looks good so far - could you provide a result for 

- [ ] Tested basic functionality of all binary files as called with `nix run $name`

**Comment by malob** [comment]
@SuperSandro2000 or @zakame, just a polite poke in case this has fallen through the cracks.

**Comment by malob** [comment]
@SuperSandro2000, anything I can do here to get this merged?

**Comment by SuperSandro2000** [comment]
I am not sure. I would appreciate @stigtsp 's feedback.

**Comment by zakame** [comment]
@malob see also #176398 from @zimbatm

**Review by stigtsp — COMMENTED** [review]
Looks sensible to me, but also not familiar with implications of adding `mainProgram`. Are there tests that check if `mainProgram` resolves/executes somewhere?

**Comment by SuperSandro2000** [comment]
> also not familiar with implications of adding `mainProgram`

Just meta information for nix run.

> Are there tests that check if mainProgram resolves/executes somewhere?

nope

File: pkgs/top-level/perl-packages.nix [hunk 2]
```diff
@@ -13943,17 +14074,18 @@ let
...
+    src = fetchurl {
+      url = "mirror://cpan/authors/id/T/TI/TINITA/Mo-0.40.tar.gz";
+      sha256 = "1fff81awg9agfawf3wxx0gpf6vgav8w920rmxsbjg30z75943lli";
+    };
+    meta = {
+      description = "Micro Objects. Mo is less.";
+      homepage = "https://github.com/ingydotnet/mo-pm";
+      license = with lib.licenses; [ artistic1 gpl1Plus ];
+      mainProgram = "mo-inline";
+    };
```
Comment (stigtsp) [inline pkgs/top-level/perl-packages.nix:14088]: Unrelated indentation changes

File: pkgs/top-level/perl-packages.nix [hunk 3]
```diff
@@ -17391,19 +17547,20 @@ let
...
+      sha256 = "021ink414w4mdk6rd54cc1f23kfqg0zk4njx4ngr0bw3wc6r4kks";
+    };
+    propagatedBuildInputs = [ FileHomeDir HTTPDate MooXOptions TimeDurationParse ];
+    buildInputs = [ PathTiny ];
+    meta = {
+      description = "interface to PAUSE's module permissions file (06perms.txt)";
+      homepage = "https://github.com/neilb/PAUSE-Permissions";
+      license = with lib.licenses; [ artistic1 gpl1Plus ];
+      mainProgram = "pause-permissions";
+    };
```
Comment (stigtsp) [inline pkgs/top-level/perl-packages.nix:17563]: Unrelated indentation changes

File: pkgs/top-level/perl-packages.nix [hunk 4]
```diff
@@ -4049,18 +4080,19 @@ let
...
+      url = "mirror://cpan/authors/id/M/MI/MIYAGAWA/Corona-0.1004.tar.gz";
+      sha256 = "0g5gpma3998rn61qfjv5csv2nrdi4sc84ipkb4k6synyhfgd3xgz";
+    };
+    propagatedBuildInputs = [ NetServerCoro Plack ];
+    buildInputs = [ TestSharedFork TestTCP ];
+    meta = {
+      description = "Coro based PSGI web server";
+      license = with lib.licenses; [ artistic1 gpl1Plus ];
+      mainProgram = "corona";
+    };
```
Comment (stigtsp) [inline pkgs/top-level/perl-packages.nix:4095]: Unrelated indentation change

File: pkgs/top-level/perl-packages.nix [hunk 5]
```diff
@@ -12995,12 +13117,13 @@ let
       url = "mirror://cpan/authors/id/W/WY/WYANT/Mac-Pasteboard-0.011.tar.gz";
       sha256 = "1a82pacp6pph3y2agdihzr4vc0phx85mq5am9czc81g8n484b35x";
     };
+    buildInputs = [ pkgs.darwin.apple_sdk.frameworks.ApplicationServices ];
```
Comment (stigtsp) [inline pkgs/top-level/perl-packages.nix:13120]: Unrelated fix for darwin build?

File: pkgs/top-level/perl-packages.nix [hunk 6]
```diff
@@ -18888,23 +19078,25 @@ let
...
+      url = "mirror://cpan/authors/id/M/MI/MINIMAL/Pod-Markdown-Github-0.04.tar.gz";
+      sha256 = "04y67c50hpf1vb9cwsza3fbj4rshdqa47vi3zcj4kkjckh02yzmk";
+    };
+    propagatedBuildInputs = [ PodMarkdown ];
+    buildInputs = [ TestDifferences ];
+    meta = {
+      description = "Convert POD to Github's specific markdown";
+      license = with lib.licenses; [ artistic1 gpl1Plus ];
+      mainProgram = "pod2github";
+    };
```
Comment (stigtsp) [inline pkgs/top-level/perl-packages.nix:19099]: Unrelated indentation change

File: pkgs/top-level/perl-packages.nix [hunk 7]
```diff
@@ -18941,18 +19138,19 @@ let
...
+      url = "mirror://cpan/authors/id/J/JH/JHOBLITT/Pod-Tidy-0.10.tar.gz";
+      sha256 = "1gcxjplgksnc5iggi8dzbkbkcryii5wjhypd7fs3kmbwx91y2vl8";
+    };
+    propagatedBuildInputs = [ EncodeNewlines IOString PodWrap TextGlob ];
+    buildInputs = [ TestCmd ];
+    meta = {
+      description = "a reformatting Pod Processor";
+      license = with lib.licenses; [ artistic1 gpl1Plus ];
+      mainProgram = "podtidy";
+    };
```
Comment (stigtsp) [inline pkgs/top-level/perl-packages.nix:19153]: Unrelated indentation change

File: pkgs/top-level/perl-packages.nix [hunk 8]
```diff
@@ -18972,16 +19170,17 @@ let
...
+    version = "0.01";
+    src = fetchurl {
+      url = "mirror://cpan/authors/id/N/NU/NUFFIN/Pod-Wrap-0.01.tar.gz";
+      sha256 = "0qwb5hp26f85xnb3zivf8ccfdplabiyl5sd53c6wgdgvzzicpjjh";
+    };
+    propagatedBuildInputs = [ PodParser ];
+    meta = {
+      license = with lib.licenses; [ artistic1 gpl1Plus ];
+      mainProgram = "podwrap";
+    };
```
Comment (stigtsp) [inline pkgs/top-level/perl-packages.nix:19183]: Unrelated indentation change

File: pkgs/top-level/perl-packages.nix [hunk 9]
```diff
@@ -19675,29 +19887,31 @@ let
...
+    version = "1.2";
+    src = fetchurl {
+      url = "mirror://cpan/authors/id/A/AS/ASKSH/Snowball-Norwegian-1.2.tar.gz";
+      sha256 = "0675v45bbsh7vr7kpf36xs2q79g02iq1kmfw22h20xdk4rzqvkqx";
+    };
+    meta = {
+      description = "Porters stemming algorithm for norwegian.";
+      license = with lib.licenses; [ artistic1 gpl1Plus ];
+      mainProgram = "stemmer-no.pl";
+    };
```
Comment (stigtsp) [inline pkgs/top-level/perl-packages.nix:19900]: Unrelated indentation change

File: pkgs/top-level/perl-packages.nix [hunk 10]
```diff
@@ -19675,29 +19887,31 @@ let
...
+    version = "1.2";
+    src = fetchurl {
+      url = "mirror://cpan/authors/id/A/AS/ASKSH/Snowball-Swedish-1.2.tar.gz";
+      sha256 = "0agwc12jk5kmabnpsplw3wf4ii5w1zb159cpin44x3srb0sr5apg";
+    };
+    meta = {
+      description = "Porters stemming algorithm for swedish.";
+      license = with lib.licenses; [ artistic1 gpl1Plus ];
+      mainProgram = "stemmer-se.pl";
+    };
```
Comment (stigtsp) [inline pkgs/top-level/perl-packages.nix:19914]: Unrelated indentation change

File: pkgs/top-level/perl-packages.nix [hunk 11]
```diff
@@ -21975,19 +22207,20 @@ let
...
+      sha256 = "18s3c8qfr3kmmyxmsn5la2zgbdsgpnkmscnl68i7fnavfpfnqlxl";
+    };
+    propagatedBuildInputs = [ ModuleCPANTSAnalyse ];
+    buildInputs = [ CPANMetaCheck TestDeep TestWarnings ];
+    meta = {
+      description = "Test the Kwalitee of a distribution before you release it";
+      homepage = "https://github.com/karenetheridge/Test-Kwalitee";
+      license = with lib.licenses; [ artistic1 gpl1Plus ];
+      mainProgram = "kwalitee-metrics";
+    };
```
Comment (stigtsp) [inline pkgs/top-level/perl-packages.nix:22223]: Unrelated indentation change

File: pkgs/top-level/perl-packages.nix [hunk 12]
```diff
@@ -24147,19 +24394,20 @@ let
...
+      sha256 = "1a57knbwync7rlzhsz1kdc0sd380xnaccwgiy1qwj5d87abdynnp";
+    };
+    propagatedBuildInputs = [ AnyEvent Plack ];
+    buildInputs = [ TestRequires TestSharedFork TestTCP ];
+    meta = {
+      description = "AnyEvent HTTP server for PSGI (like Thin)";
+      homepage = "https://github.com/miyagawa/Twiggy";
+      license = with lib.licenses; [ artistic1 gpl1Plus ];
+      mainProgram = "twiggy";
+    };
```
Comment (stigtsp) [inline pkgs/top-level/perl-packages.nix:24410]: Unrelated indentaiton change

**Review by stigtsp — CHANGES_REQUESTED** [review]
Hi there, the `mainProgram` changes LGTM. 

However, I've had a look at the diff and found a few unrelated changes in `perl-packages.nix`. Mainly:
- Indentation changes
- Changed order of some attributes
- ~Changes to dependencies for `MacPasteboard`~ Was confused by attribute order change :confused: 
- Additional meta (no biggie)

The indentation changes makes it a bit difficult to review what has actually changed, and could cause merge problems imho.

File: pkgs/top-level/perl-packages.nix [hunk 13]
```diff
@@ -5124,6 +5165,7 @@ let
     meta = {
       description = "Hexadecimal Dumper";
       maintainers = with maintainers; [ AndersonTorres ];
+      mainProgram = "hexdump";
```
[the remark this answers is not in the document]
Comment (SuperSandro2000) [reply pkgs/top-level/perl-packages.nix:5168]: I don't think there is a concept of conflicts for mainProgram.

[the remark this answers is not in the document]
Comment (stigtsp) [reply pkgs/top-level/perl-packages.nix:5168]: Ah ofc not, thx :)

**Comment by malob** [comment]
Alrighty I've,

* rebased on `master`,
* removed redundant `mainProgram` definitions due to #176398, and
* moved all indentation changes to a 3rd commit which fixes indentation issues in the whole file.

@stigtsp, sorry if the indentation changes and the reshuffling of attributes made the commits harder to review. I have (what I think is generally a good) habit of making small fixes/improvements unrelated to the changes I want to make when I notice opportunities to do so. In many cases these changes are minor enough that they don't warrant a separate commit, or like in the case of reshuffling the attributes, I wouldn't bother to make those improvements if I had to put them all into a separate commit (since it's annoying to note them all down, and come back to them separately). 

So for changes like the shuffling of attributes, I usually lean towards including them, even if it makes the commit a little harder to review, since the alternative of in my case would be to just leave those changes off (which seems worse, at least to me, overall).

However, I agree with you that the indentation changes in particular definitely deserved a separate commit since the indentation errors are easy to find on correct on their own, and they made reviewing the other changes a lot more annoying.

**Review by zimbatm — APPROVED** [review]

**Comment by malob** [comment]
*Politely poking @stigtsp and/or @SuperSandro2000*

**Review by SuperSandro2000 — APPROVED** [review]
